src/Barbies/Internal/DistributiveB.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE PolyKinds    #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+module Barbies.Internal.DistributiveB
+  ( DistributiveB(..)
+  , bshape
+  , bdistribute'
+  , gbdistributeDefault
+  , CanDeriveDistributiveB
+  )
+
+where
+
+import Barbies.Internal.FunctorB (FunctorB(..))
+import Barbies.Generics.Distributive (GDistributive(..))
+
+import Data.Functor.Compose   (Compose (..))
+import Data.Functor.Identity  (Identity (..))
+import Data.Functor.Product   (Product (..))
+import Data.Generics.GenericN
+import Data.Proxy             (Proxy (..))
+import Data.Distributive
+import Data.Kind              (Type)
+
+-- | A 'FunctorB' where the effects can be distributed to the fields:
+-- `bdistribute` turns an effectful way of making a Barbie-type
+-- into a pure Barbie-type with effectful ways of computing the
+-- values of its fields.
+--
+-- From 'bdistribute', we can define a pure skeleton where very field
+-- is a getter of that field:
+--
+-- @
+-- 'bshape' :: ('DistributiveB' b) => b ((->) (b 'Identity'))
+-- 'bshape' = 'bdistribute'' 'id'
+-- @
+--
+--  It should satisfy the following law
+--
+-- @
+-- 'bmap' ('Identity' . ('$' b)) 'bshape' = b
+-- @
+--
+--  It is to 'FunctorB' in the same way as 'Distributive'
+--  relates to 'Functor'.
+--
+-- There is a default implementation of 'bdistribute' based on
+-- 'Generic'.  Intuitively, it works on product types where the shape
+-- of a pure value is uniquely defined and every field is covered by
+-- the argument @f@.
+class (FunctorB b) => DistributiveB (b :: (k -> Type) -> Type) where
+  bdistribute :: Distributive f => f (b g) -> b (Compose f g)
+  -- bdistribute x = bmap (\f -> Compose $ fmap f . bsequence' <$> x) bshape
+
+  default bdistribute
+    :: forall f g
+    .  CanDeriveDistributiveB b f g
+    => Distributive f => f (b g) -> b (Compose f g)
+  bdistribute = gbdistributeDefault
+
+bshape :: (DistributiveB b) => b ((->) (b Identity))
+bshape = bdistribute' id
+
+-- | A version of `bdistribute` with @g@ specialized to `Identity`.
+bdistribute' :: (DistributiveB b, Distributive f) => f (b Identity) -> b f
+bdistribute' = bmap (fmap runIdentity . getCompose) . bdistribute
+
```

Comment:
Maybe add `bcollect` as well as a function?

Reply:
As a function, or as a method in `DistributiveB`?

Reply:
I'd say as a function. Once you introduce default method definitions using generics, it's not that convenient to have multiple minimal definitions.